doc/doxygen/src/driver-guide.md

```diff
@@ -0,0 +1,352 @@
+great care should be put into ...(?)
+
+Third, it should always be possible, to handle more than a single device of one
+kind. Drivers and their interfaces are thus designed to keep their state
+information in a parameterized location instead of driver defined global
+variables.
+
+Fourth, RIOT defines high-level interfaces for certain groups of devices (i.e.
+netdev for network devices, SAUL for sensors and actuators), which enables users
+to work with a wide variety of devices without having to know anything about the
+actual device that is mapped.
+
+Fifth, during initialization we make sure that we can communicate with a device.
+Other functions should check the dev pointer is not void, and should also handle
+error return values from the lower layer peripheral driver implementations,
+where there are some.
+
+Sixth, device drivers SHOULD be implemented independent of any CPU/board code.
+To achieve this, the driver implementations should be strictly be based on
+platform independent interfaces as the peripheral drivers, xtimer, etc.
+
+
+# General                                               {#driver-guide-general}
+
+## Documentation                                            {#driver-guide-doc}
+Document what your driver does! Most device come with a very large number of
+features, while the corresponding device driver only supports a sub-set of them.
+This should be clearly stated in the device driver's documentation, so that
+anyone wanting to use the driver can find out the supported features without
+having to scan through the code.
+
+## Device descriptor and parameter configuration          {#driver-guide-types}
+Each device MUST supply a data structure, holding the devices state and
+configuration, using the naming scheme of `DEVNAME_t` (e.g. `dht_t`, or
+`at86rf2xx_t`). In context of RIOT, we call this data structure the device
+descriptor.
+
+This device descriptor MUST contain all the state data of a device. By this, we
+are not limited on the number of instances of the driver we can run
+concurrently. The descriptor is hereby used for identifying the device we want
+to interact with, and SHOULD always be the first parameter for all device driver
+related functions.
+
+Typical things found in these descriptors are e.g. used peripherals (e.g. SPI or
+I2C bus used, interfacing GPIO pins), data buffers (e.g. RX/TX buffers where
+needed), or state machine information.
+
+On top of the device descriptor, each device driver MUST also define a data
+structure holding the needed configuration data. The naming scheme for this type
+is `DEVNAME_params_t`. In contrary to the device descriptor, this data structure
+should only contain static information, that is needed for the device
+initialization as it is preferably allocated in ROM.
+
+A simple I2C temperature sensors's device descriptor could look like this:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
```

Comment:
Something still isn't right... I think you need to remove the whitespace

```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.c}
```

However, while enabling syntax highlighting with this, the latter code blocks are still not rendered properly...

Reply:
For the record, this is how the page currently renders

![image](https://user-images.githubusercontent.com/675644/69536855-20932980-0f7f-11ea-8070-d31fdfcf81b8.png)

Reply:
On my system it worked fine. I used a markdown linter to verify the syntax now. It complained about missing blank lines before or after then fenced code blocks. Maybe this is what the old version of the Doxygen didn't handle correctly. Let's see, if this now works.

(However: A numbered list containing fenced code blocks will not work, when having blank lines in between. I had to change the source to `First, <first item>. Second, <second item>, ...`.)

Reply:
Maybe its the `...` that the syntax highlighter in the code blocks cannot handle? I changed those to `/* ... */` to create valid c syntax.

Reply:
> (However: A numbered list containing fenced code blocks will not work, when having blank lines in between. I had to change the source to `First, <first item>. Second, <second item>, ...`.)

Did you try

```markdown
1. Blafoo

   ~~~~~~~~~ {.c}
   /* some code */
   ~~~~~~~~~
```
?

Reply:
*Still* doesn't build with 1.8.13 btw :confused:

Reply:
> Did you try

Just did, but with that I also get incorrectly rendered markdown locally (with the current Doxygen version).

Reply:
Using `@code` and `@endcode` instead of the fences works reliable on the old version of Doxygen. But that still doesn't work in conjunction with the numbered list.